modal/image.py

```diff
@@ -682,9 +682,10 @@ def _add_local_python_packages(self, *packages: Union[str, Path], copy: bool = F
         required if you want to run additional build steps after this one.
 
         **Note:** This excludes all dot-prefixed subdirectories or files and all `.pyc`/`__pycache__` files.
-        To add full directories with finer control, use `.add_local_dir()` instead.
```

Comment:
I still wonder if we should change the default condition here to include only `.py` files, or maybe only directories containing `.py` files or something.

Reply:
I'm open to it! We need to add easy to override filters for users though - @op just ran into issues with this and resorted back to `copy_mount(Mount...)` to get around the limitations :(

Reply:
I used something like this:

```python
def allow_globs(*patterns: str):
    def condition(file_path: str) -> bool:
        path = Path(file_path).relative_to(repo_root)
        for pattern in patterns:
            if fnmatch(str(path), pattern):
                return True
        return False

    return condition


modal_mount = Mount.from_local_dir(
    ".",
    condition=allow_globs(
        "Cargo.toml",
        "ffi/**/*.rs",
        "crates/**/*.rs",
        "crates/**/*.proto",
        "crates/**/*.sql",
        "**/Cargo.toml",
        "**/requirements.dev.txt",
    ),
)
```

Reply:
Don't forget that we now also have a dockerignore-compliant pattern matcher in our library: https://github.com/modal-labs/modal-client/blob/main/modal/_utils/pattern_matcher.py

Reply:
Oh wow I wasn't aware of this! @kramstrom this sounds very relevant to the stuff you're working on!

Reply:
Yep we'd just held off on figuring out exactly what the API should be (i.e. are we using it for exclusion or inclusion or both? do you pass a list of string patterns or a file? etc.)

Reply:
I just pitched to Kasper that maybe the easiest for users would to have it always specify inclusion, but have "set arithmetic" operators (`+` and `-` for simplicity) that lets you combine them in any way you please in order to build complex include/excludes.

E.g.

`expr = files("**/*.py", "*.data") - files("bigblob.data")`